tools/python/mwm_downloader.py

```diff
@@ -0,0 +1,152 @@
+import logging
+
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+
+DIRECT_MAP_URL = 'http://direct.mapswithme.com/direct/'
+ROOT = os.path.dirname(os.path.realpath(__file__))
+DEFAULT_DOWNLOAD_DIRECTORY = os.path.join(os.path.dirname(os.path.dirname(ROOT)), 'data')
+MAP_REVISION_REGEXP = re.compile(r'(\d{6})')
+MWM_NAME_REGEXP = re.compile(r'"(\S+\.mwm)"')
+
+logger = logging.getLogger(__name__)
+
+
+def parse_html_by_regexp(url, regexp):
+    def parse_line(line):
+        return regexp.search(line.decode('utf-8'))
+
+    with urllib.request.urlopen(url) as f:
+        lines = f.readlines()
+        filtered_lines = filter(None, map(parse_line, lines))
+        return (line.group(1) for line in filtered_lines)
+
+
+def get_mwm_versions():
+    return map(int, parse_html_by_regexp(DIRECT_MAP_URL, MAP_REVISION_REGEXP))
+
+
+def get_mwm_names(version_num):
+    version_url = '{base}{version}'.format(base=DIRECT_MAP_URL, version=version_num)
+    return map(urllib.parse.unquote, parse_html_by_regexp(version_url, MWM_NAME_REGEXP))
+
+
+def download_file(url, filename, attempts=3):
+    try:
+        urllib.request.urlretrieve(url, filename=filename)
+    except (urllib.error.URLError, PermissionError):
+        logger.error('Could not download map {name}'.format(name=os.path.basename(filename)), exc_info=True)
+        if attempts:
+            time.sleep(3)
+            return download_file(url, filename, attempts-1)
+        return False
+    return True
+
+
+def progress_bar(total, progress):
+    """
+    Displays or updates a console progress bar.
+
+    Original source: https://stackoverflow.com/a/15860757/1391441
+    """
+    initial_progress = progress
+    bar_length = 20
+    progress = progress / total
+    block = int(round(bar_length * progress))
+    progress_bar_str = '#' * block + '-' * (bar_length - block)
+    text = '\r[{progress_bar}] {percent:.0f}% {item}/{total}'.format(
+        progress_bar=progress_bar_str,
+        percent=round(progress * 100, 0),
+        item=initial_progress,
+        total=total)
+    sys.stderr.write(text)
+    if progress == 1:
+        sys.stderr.write('\n')
+    sys.stderr.flush()
+
+
+def download_mwm_list(version, threads, folder, mwm_prefix_list=None, quiet=True):
+    versions = list(get_mwm_versions())
+
+    if version is None:
+        # Get latest version by default
+        version = max(versions)
+    elif version not in versions:
+        logger.error('Maps with version {version} is not found in {url}'.format(version=version, url=DIRECT_MAP_URL))
+        return
+
+    mwm_names = list(get_mwm_names(version))
+    mwm_to_download = []
+    if mwm_prefix_list is not None:
+        for prefix in mwm_prefix_list:
+            filtered_mwm = [mwm_name for mwm_name in mwm_names if mwm_name.startswith(prefix)]
+            if not filtered_mwm:
+                logger.error('No map with prefix \'{prefix}\' was found in {url}'.format(prefix=prefix, url=DIRECT_MAP_URL))
```

Comment:
if you can ignore faulty prefixes you can write this as

prefixes = re.compile('|'.join(mwm_prefix_list))
mwm_to_download = filter(lambda n: prefixes.match(n), mwm_names)

Reply:
Thanks, looks nice! But i've replaced ```mwm_prefix_list``` with: 
```python3
('^{prefix}' for prefix in mwm_prefix_list)
``` 
to avoid unexpected matches in the middle of the map name.
Fixed.